### PR TITLE
[Merged by Bors] - feat: interaction of `Finset.sup` and `Nat.cast`

### PR DESCRIPTION
**Pull request by YaelDillies**
Co-authored-by: Andrew Yang

---

- [x] depends on: #19461

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: scripts/no_lints_prime_decls.txt [hunk 1]
```diff
@@ -3228,6 +3228,8 @@ Nat.add_descFactorial_eq_ascFactorial'
 Nat.ascFactorial_eq_factorial_mul_choose'
 Nat.bit_add'
 Nat.card_eq_two_iff'
+Nat.cast_finsetInf'
+Nat.cast_finsetSup'
```
Comment (Ruben-VandeVelde) [inline scripts/no_lints_prime_decls.txt:3232]: Those are about `inf'`/`sup'` so should probably be nolinted at the declaration

File: Mathlib/Algebra/Order/Group/Finset.lean [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+namespace Finset
+variable {ι κ M G : Type*}
+
+lemma fold_max_add [LinearOrder M] [Add M] [AddRightMono M] (s : Finset ι) (a : WithBot M)
+    (f : ι → M) : s.fold max ⊥ (fun i ↦ ↑(f i) + a) = s.fold max ⊥ ((↑) ∘ f) + a := by
+  classical
+    induction' s using Finset.induction_on with a s _ ih <;> simp [*, max_add_add_right]
+
+@[to_additive nsmul_inf']
+lemma inf'_pow [LinearOrder M] [Monoid M] [MulLeftMono M] [MulRightMono M] (s : Finset ι)
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/Order/Group/Finset.lean:26]: Probably `s` should be implicit, and I think in all the other new lemmas as well?

File: Mathlib/Algebra/Order/Group/Finset.lean [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    (f : ι → M) (n : ℕ) (hs : s.Nonempty) : s.inf' hs f ^ n = s.inf' hs fun a ↦ f a ^ n :=
+  let ns : InfHom M M := { toFun := (· ^ n), map_inf' := fun _ _ ↦ (pow_left_mono n).map_min }
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/Order/Group/Finset.lean:28]: Is it worth making this a definition? (Can be deferred)

File: Mathlib/Algebra/Order/Group/Finset.lean [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+import Mathlib.Algebra.Order.Monoid.Unbundled.MinMax
+import Mathlib.Algebra.Order.Monoid.Unbundled.Pow
+import Mathlib.Algebra.Order.Monoid.Unbundled.WithTop
+import Mathlib.Data.Finset.Lattice.Fold
+
+/-!
+# Arithmetic of `Finset.sup`
+-/
+
+namespace Finset
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/Order/Group/Finset.lean:17]: Consider `assert_not_exists GroupWithZero`

**Review by Ruben-VandeVelde — APPROVED** [review]

File: Mathlib/Algebra/Order/Group/Finset.lean (hunk 2 above)
Comment (YaelDillies) [reply Mathlib/Algebra/Order/Group/Finset.lean:26]: This is an exception to the rule that arguments that can be inferred from later ones should be implicit. The reason is that `hs` can be filled in by unification from both the LHS and RHS when rewriting, so it won't need to be provided. In contrast, the `s` argument can be used to disambiguate rewrites

File: Mathlib/Algebra/Order/Group/Finset.lean (hunk 3 above)
Comment (YaelDillies) [reply Mathlib/Algebra/Order/Group/Finset.lean:28]: Don't think so. I have inlined it

File: Mathlib/Algebra/Order/Group/Finset.lean (hunk 4 above)
Comment (YaelDillies) [reply Mathlib/Algebra/Order/Group/Finset.lean:17]: Well spotted

File: Mathlib/Algebra/Order/Ring/Finset.lean [hunk 5]
```diff
@@ -1,17 +1,21 @@
...
 
 namespace Finset
-variable {ι R : Type*} [LinearOrderedSemiring R] {a b : ι → R}
+variable {ι R : Type*}
+
+section LinearOrderedSemiring
+variable [LinearOrderedSemiring R] {s : Finset ι} {a b : ι → R}
 
 theorem sup_mul_le_mul_sup_of_nonneg [OrderBot R] (s : Finset ι) (ha : ∀ i ∈ s, 0 ≤ a i)
     (hb : ∀ i ∈ s, 0 ≤ b i) : s.sup (a * b) ≤ s.sup a * s.sup b :=
```
Comment (b-mehta) [inline Mathlib/Algebra/Order/Ring/Finset.lean:21]: Don't these lemmas also work for something much weaker than semiring? It looks like this one needs less than `sup'_mul₀`.

File: Mathlib/Algebra/Order/GroupWithZero/Finset.lean [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+# Arithmetic of `Finset.sup`
+-/
+
+namespace Finset
+variable {ι G₀ : Type*}
+
+section GroupWithZero
+variable [GroupWithZero G₀] [SemilatticeSup G₀] {s : Finset ι} {a : G₀}
+
+lemma sup'_mul₀ [MulPosMono G₀] [MulPosReflectLE G₀] (ha : 0 < a) (f : ι → G₀) (s : Finset ι) (hs) :
```
[the remark this answers is not in the document]
Comment (YaelDillies) [reply Mathlib/Algebra/Order/GroupWithZero/Finset.lean:20]: Added the other side version of this one. I doubt the other side version of the `div` is true

File: Mathlib/Algebra/Order/Group/Finset.lean [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+-/
+import Mathlib.Algebra.Order.Group.OrderIso
+import Mathlib.Algebra.Order.Monoid.Canonical.Defs
+import Mathlib.Algebra.Order.Monoid.Unbundled.MinMax
+import Mathlib.Algebra.Order.Monoid.Unbundled.Pow
+import Mathlib.Algebra.Order.Monoid.Unbundled.WithTop
+import Mathlib.Data.Finset.Lattice.Fold
+
+/-!
+# Arithmetic of `Finset.sup`
```
[the remark this answers is not in the document]
Comment (YaelDillies) [reply Mathlib/Algebra/Order/Group/Finset.lean]: What about now?

File: Mathlib/Algebra/Order/Group/Finset.lean [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+lemma sup'_add' (s : Finset ι) (f : ι → M) (a : M) (hs : s.Nonempty) :
+    s.sup' hs f + a = s.sup' hs fun i ↦ f i + a := by
+  apply le_antisymm
+  · apply add_le_of_le_tsub_right_of_le
+    · exact Finset.le_sup'_of_le _ hs.choose_spec le_add_self
+    · exact Finset.sup'_le _ _ fun i hi ↦ le_tsub_of_add_le_right (Finset.le_sup' (f · + a) hi)
+  · exact Finset.sup'_le _ _ fun i hi ↦ add_le_add_right (Finset.le_sup' _ hi) _
+
+/-- Also see `Finset.add_sup'` that works for canonically ordered monoids. -/
+lemma add_sup'' (hs : s.Nonempty) (f : ι → M) (a : M) :
```
[the remark this answers is not in the document]
Comment (YaelDillies) [reply Mathlib/Algebra/Order/Group/Finset.lean:65]: I agree :/

File: Mathlib/Algebra/Order/Group/Finset.lean [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+    (f : ι → M) (n : ℕ) (hs) : s.sup' hs f ^ n = s.sup' hs fun a ↦ f a ^ n :=
+  map_finset_sup' (OrderHom.mk _ <| pow_left_mono n) hs _
+
+section Group
+variable [Group G] [LinearOrder G]
+
+/-- Also see `Finset.sup'_mul'` that works for canonically ordered monoids. -/
+@[to_additive "Also see `Finset.sup'_add'` that works for canonically ordered monoids."]
+lemma sup'_mul [MulRightMono G] (s : Finset ι) (f : ι → G) (a : G) (hs) :
+    s.sup' hs f * a = s.sup' hs fun i ↦ f i * a := map_finset_sup' (OrderIso.mulRight a) hs f
```
[the remark this answers is not in the document]
Comment (b-mehta) [reply Mathlib/Algebra/Order/Group/Finset.lean]: Interesting, I have exactly the opposite experience, that further simplifications are available when keeping the multiplication on the outside. For example, `s.sup' hs f - a = b` allows algebraic manipulations involving the entire equation, and these usually mean the goal is easier to prove.

Similarly when writing informally, we typically keep as little inside a supremum as possible: `Sup_i (f i + c)` is almost always viewed as less simple than `Sup_i (f i) + c`.

But let's see how these ones get used. Certainly I'll be almost exclusively using them in reverse :)

File: Mathlib/Algebra/Order/Ring/Finset.lean [hunk 10]
```diff
@@ -1,34 +1,48 @@
...
+Authors: Eric Wieser, Yaël Dillies, Andrew Yang
 -/
-import Mathlib.Algebra.Order.Ring.Defs
+import Mathlib.Algebra.Order.Field.Canonical.Defs
+import Mathlib.Algebra.Order.Floor
 import Mathlib.Data.Finset.Lattice.Fold
 
 /-!
-# Algebraic properties of finitary supremum
+# `Finset.sup` in a ring
```
Comment (b-mehta) [inline Mathlib/Algebra/Order/Ring/Finset.lean:11]: Now this file is just about casts interacting with `Finset.sup`, so maybe a docstring describing it that way would be nice

**Review by b-mehta — APPROVED** [review]
~~It looks like the `Group/Finset` lemmas went missing?~~ oops, now I see!

File: Mathlib/Algebra/Order/Ring/Finset.lean [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+
+set_option linter.docPrime false in
+@[simp, norm_cast]
+lemma cast_finsetInf' (f : ι → ℕ) (hs) : (↑(s.inf' hs f) : R) = s.inf' hs fun i ↦ (f i : R) :=
+  eq_of_forall_le_iff fun c ↦ by simp only [le_inf'_iff, ← Nat.ceil_le]
+
+end LinearOrderedSemiring
+
+@[simp, norm_cast]
+lemma cast_finsetSup [CanonicallyLinearOrderedSemifield R] [FloorSemiring R]
```
Comment (b-mehta) [inline Mathlib/Algebra/Order/Ring/Finset.lean:42]: following the above, this one doesn't need floor semiring any more.

File: Mathlib/Algebra/Order/Ring/Finset.lean [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+      apply le_floor
+      exact le_sup' (fun i ↦ (f i : R)) hi
+    · exact le_sup'_of_le (fun i ↦ (f i : R)) hs.choose_spec (Nat.cast_nonneg _)
+  apply sup'_le
+  exact fun i hi ↦ Nat.mono_cast (le_sup' _ hi)
+
+set_option linter.docPrime false in
+@[simp, norm_cast]
+lemma cast_finsetInf' (f : ι → ℕ) (hs) : (↑(s.inf' hs f) : R) = s.inf' hs fun i ↦ (f i : R) :=
+  eq_of_forall_le_iff fun c ↦ by simp only [le_inf'_iff, ← Nat.ceil_le]
```
[the remark this answers is not in the document]
Comment (YaelDillies) [reply Mathlib/Algebra/Order/Ring/Finset.lean:37]: Oh nice. I didn't think an adjoint-less proof was possible

File: Mathlib/Algebra/Order/Ring/Finset.lean [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+
+end LinearOrderedSemiring
+
+@[simp, norm_cast]
+lemma cast_finsetSup [CanonicallyLinearOrderedSemifield R] [FloorSemiring R]
+    (s : Finset ι) (f : ι → ℕ) : (↑(s.sup f) : R) = s.sup fun i ↦ (f i : R) := by
+  obtain rfl | hs := s.eq_empty_or_nonempty
+  · simp
+  · rw [← sup'_eq_sup hs, ← sup'_eq_sup hs, Nat.cast_finsetSup']
```
[the remark this answers is not in the document]
Comment (b-mehta) [reply Mathlib/Algebra/Order/Ring/Finset.lean]: btw, @YaelDillies, I think we should have the version of `comp_sup_eq_sup_comp` that works for `sup'`, asks for the finset to be nonempty and has no assumption about what happens to bot. Its proof should be analogous to the ones I gave above, but it gives those two lemmas as easy special cases

[the remark this answers is not in the document]
Comment (b-mehta) [reply Mathlib/Algebra/Order/Ring/Finset.lean]: Oh! Then I'm still correct that it should exist ;)

**Review by jcommelin — COMMENTED** [review]
bors merge

**Pull request closed without merging**